### ISSUE TITLE
lint: emit json

### ISSUE THREAD

**Issue by dominikh** [dominikh/go-tools#21]
We should provide a flag to output results in JSON. Such output would allow other tools to more easily consume our output.

**Comment by shurcooL** [comment]
When you say other tools, do you mean tools that do not have a good way of parsing the `file:line:col: text` type of output?

**Comment by dominikh** [comment]
Well, tools like frontends shouldn't have to parse that format (which currently exists of 3 components: the location, the message, and the ID of the check). It'd be much cleaner if they could receive the information in a structured format.

**Comment by dominikh** [comment]
Preliminary design of the JSON output is a stream of JSON objects of the following kind:

```
{
  "checker": "staticcheck",
  "code": "SA1000",
  "severity": "warning",
  "location": {
    "file": "/some/absolute/path.go",
    "line": 123,
    "column": 123,
  },
  "message": "Your code is full of bugs"
}
```

Where `severity` is optional, and not fully defined yet, but will be used to differentiate compiler errors, suggestions (gosimple), warnings (most of staticcheck), high impact bugs and so on. We can standardize on a list later.

`column` is a byte offset, just like it's in the existing text output.

@bradleyfalzon as the author of gopherci, do you have any feedback on the format? any fields that are missing?

**Comment by dominikh** [comment]
Real-world example (without the `severity` field):

```
megacheck strings
{"checker":"staticcheck","code":"SA4017","location":{"file":"/usr/lib/go/src/strings/strings_test.go","line":954,"column":8},"Message":"Repeat is a pure function but its return value is ignored"}
{"checker":"gosimple","code":"S1021","location":{"file":"/usr/lib/go/src/strings/replace_test.go","line":54,"column":2},"Message":"should merge variable declaration with assignment on next line"}
```

**Comment by bradleyfalzon** [comment]
Looks good to me, not sure if others would prefer to handle a single json document instead of a stream. 

What about errors, or is it just assumed that if you weren’t able to decode a plain text error is presented itself?

**Comment by dominikh** [comment]
> not sure if others would prefer to handle a single json document instead of a stream.

It's much easier to be given a stream and treat it as a single document (read until EOF before processing) than it is to do streaming with a document.

> What about errors, or is it just assumed that if you weren’t able to decode a plain text error is presented itself?

For now, if the input is malformed, we'll continue emitting errors on stderr. In the future, these may become part of the JSON.

**Comment by dominikh** [comment]
Implemented in 479b882bdda7ac0c9d3e292e59a59bf15dcdcb71
